stdlib/public/core/NFD.swift

```diff
@@ -0,0 +1,220 @@
+      }
+
+      buffer.append((scalar: scalar, normData: normData))
+    }
+  }
+
+  internal mutating func next() -> (scalar: Unicode.Scalar, normData: UInt16)? {
+    // Empty out our buffer before attempting to decompose the next
+    // normalization segment.
+    if !buffer.isEmpty {
+      return buffer.removeLast()
+    }
+    
+    // If we have a starter stored, take care of that before asking for more
+    // scalars.
+    if let starter = starter.take() {
+      let ccc = starter.normData >> 3
+      var isNFDQC = starter.normData & 1 == 0
+      
+      // Because we don't store precomposed hangul in our NFD_QC data, these
+      // will return true for NFD_QC when in fact they are not.
+      if (0xAC00 ... 0xD7A3).contains(starter.scalar.value) {
+        isNFDQC = false
+      }
+      
+      // If our scalar IS NFD quick check, then it's as simple as appending to
+      // our buffer and moving on the next scalar. Otherwise, we need to
+      // decompose this and append each decomposed scalar.
+      if isNFDQC {
+        // Fast path: If our scalar is also ccc = 0, then this doesn't need to
+        // be appended to the buffer at all.
+        if ccc == 0 {
+          return starter
+        }
+
+        buffer.append(starter)
+      } else {
+        decompose(starter.scalar, with: starter.normData)
+      }
+    }
+    
+    while let scalar = iterator.next() {
+      let normData = _swift_stdlib_getNormData(scalar.value)
+      let ccc = normData >> 3
+      var isNFDQC = normData & 1 == 0
+      
+      // Because we don't store precomposed hangul in our NFD_QC data, these
+      // will return true for NFD_QC when in fact they are not.
+      if (0xAC00 ... 0xD7A3).contains(scalar.value) {
+        isNFDQC = false
+      }
+      
+      // If we've reached a starter, stop.
+      if ccc == 0, !buffer.isEmpty {
+        starter = (scalar, normData)
+        break
+      }
+      
+      // If our scalar IS NFD quick check, then it's as simple as appending to
+      // our buffer and moving on the next scalar. Otherwise, we need to
+      // decompose this and append each decomposed scalar.
+      if isNFDQC {
+        // Fast path: If our scalar is also ccc = 0, then this doesn't need to
+        // be appended to the buffer at all.
+        if ccc == 0 {
+          return (scalar, normData)
+        }
+
+        buffer.append((scalar, normData))
+      } else {
+        decompose(scalar, with: normData)
+      }
+    }
+    
+    // If we made it this far with an empty buffer, we're done.
+    if buffer.isEmpty {
+      return nil
+    }
+    
+    // Sort the entire buffer based on the canonical combining class.
+    buffer._insertionSort(within: buffer.indices) {
+      ($0.normData >> 3) < ($1.normData >> 3)
+    }
+    
+    // Reverse the buffer after sorting to make removals faster from the back
+    // rather than removing from the front.
+    buffer.reverse()
```

Comment:
Again, a read index is probably clearer. You can even make a `_NormDataBuffer` struct that holds the array and the read index and can just encapsulate the fundamental operations.